Fix AssemblyName version

sepidehMS opened the pull request:
Fixes https://github.com/dotnet/corefx/issues/19179 by using the managed CoreRT implementation of AssemblyName formatting. Only it checkes for [`canonicalizedVersion.Revision != ushort.MaxValue`](https://github.com/dotnet/coreclr/compare/master...sepidehMS:FixAssemblyFullName?expand=1#diff-fcc2014f65447c529681c711311362b9R45) and does not use RuntimeAssemblyName.

cc: @jkotas @AtsushiKan @danmosemsft

weshaggard commented on src/mscorlib/System.Private.CoreLib.csproj:
```diff
@@ -385,6 +383,8 @@
   <ItemGroup>
     <Compile Include="$(BclSourcesRoot)\System\Reflection\Assembly.CoreCLR.cs" />
     <Compile Include="$(BclSourcesRoot)\System\Reflection\AssemblyName.cs" />
+    <Compile Include="$(BclSourcesRoot)\System\Reflection\AssemblyNameHelpers.cs" />
```
We should probably move these under shared so they are correctly mirrored between coreclr and corert.

sepidehMS replied:
@weshaggard These new files live [here](https://github.com/dotnet/corert/tree/893cbdc5cf61089085ebf0d7a7a81fc54495020f/src/System.Private.CoreLib/src/System/Reflection/Runtime/Assemblies) in CoreRT. Should I keep the same structure?

weshaggard replied:
They should just get moved under shared so that they mirror between the 2.

danmosemsft replied:
yes it would be shared\system\reflection\...

When the auto-PR reflects this change into CoreRT, the PR build will break there because it will now have this type in two places. You would then go update that PR there to delete the old one to fix the PR so it can be merged.

weshaggard replied:
Or you can move them in CoreRT first and then let them mirror to CoreCLR

danmosemsft replied:
Note the files that are shared in coreclr are listed in \mscorlib\shared\System.Private.CoreLib.Shared.shproj not in thecproj.

jkotas replied:
I would recommend to not worry about the sharing for ZBB. Make this work well for CoreCLR, without sharing. We can figure out the sharing with CoreRT in separate PR - it can wait after ZBB.

AtsushiKan commented on src/mscorlib/src/System/Reflection/AssemblyNameHelpers.cs:
```diff
@@ -0,0 +1,168 @@
+            if (0 != (a.Flags & AssemblyNameFlags.Retargetable))
+                sb.Append(", Retargetable=Yes");
+
+            AssemblyContentType contentType = a.Flags.ExtractAssemblyContentType();
```
Careful now: RuntimeAssemblyName.Flags and AssemblyName.Flags don't return the same value. In particular, AssemblyName.Flags don't include the content type. You have to get that from AssemblyName.ContentType.

jkotas commented on src/mscorlib/src/System/Reflection/AssemblyNameHelpers.cs:
```diff
@@ -0,0 +1,168 @@
+                    sb.Append(", Version=");
+                    sb.Append(canonicalizedVersion.Major);
+                    sb.Append('.');
+                    sb.Append(canonicalizedVersion.Minor);
```
The check needs to be for all parts, e.g.

```
   AssemblyName assemblyName = new AssemblyName("MyAssemblyName");
   assemblyName.Version = new Version(42, 42);
   Console.WriteLine(assemblyName.FullName);
```

should print `MyAssemblyName, Version=42.42`

jkotas replied:
Actually, `sb.Append(a.Version.ToString())` may work just fine instead all this.

AtsushiKan commented on src/mscorlib/src/System/Reflection/AssemblyNameHelpers.cs:
```diff
@@ -0,0 +1,168 @@
+
+            byte[] pkt = null;
+            if (0 != (a.Flags & AssemblyNameFlags.PublicKey))
+                pkt = a.GetPublicKeyToken();
```
Note that GetPublicKeyToken() can side-effect the AssemblyName (one reason why corert uses RuntimeAssemblyName for everything - it avoids all the eccentricities of AssemblyName.)

I'm not sure if this is critical, but it is an AssemblyName that you don't own.

AtsushiKan replied:
>Note that GetPublicKeyToken() can side-effect 

After offline conversation, I'll set aside this concern for ZBB. Post-ZBB, we can do a proper sharing of this stuff that should get rid of this good.

AtsushiKan commented on src/mscorlib/src/System/Reflection/AssemblyNameLexer.cs:
```diff
@@ -0,0 +1,137 @@
+            return Token.String;
+        }
+
+        internal static KeyValuePair<char, String>[] EscapeSequences =
```
Including all of AssemblyNameLexer just to get EscapeSequences seems excesssive (unless you decide to throw the whole set into the shared partition.)

AtsushiKan commented on src/mscorlib/src/System/Reflection/AssemblyNameHelpers.cs:
```diff
@@ -0,0 +1,168 @@
+                sb.AppendQuoted(cultureName);
+            }
+
+            byte[] pkt = null;
```
The way you've changed this, if the AssemblyName doesn't have the full-length public key (which is the typical case), the generated fullname will be missing the PKT altogether.

sepidehMS replied:
@AtsushiKan Isn't that the Desktop behavior as mentioned [here](https://github.com/dotnet/corefx/issues/19179)?

AtsushiKan replied:
No, I'm talking about a case like this:

```
        public static void Main()
        {
            AssemblyName a = new AssemblyName("Foo, Version=1.0.0.0");
            a.SetPublicKeyToken(typeof(object).Assembly.GetName().GetPublicKeyToken());

            string s = a.FullName;
            Console.WriteLine(s);
        }
```

The PublicKey and PublicKeyToken are separate fields in AssemblyName. It's common to have PublicKeyToken set to an actual value while PublicKey is set to null. In such a case, the FullName is supposed to include the PublicKeyToken.

sepidehMS commented:
Addressed feedback. As @jkotas suggested, I will share the files with CoreRT in a separate PR post-ZBB, as that might introduce complications. I moved these files to a separate directory, though. PTAL. Thanks.

danmosemsft commented:
Please also take a look at our test coverage in corefx.

sepidehMS commented:
@dotnet-bot test Tizen armel Cross Debug Build please

sepidehMS commented:
@danmosemsft I checked and ran the tests locally. Will modify the tests and re-enable them once this gets merged and is picked up in CoreFX.

danmosemsft commented:
@sepidehms yeah, I meant more shredding we have sufficient coverage for this new code...

sepidehMS commented:
@dotnet-bot test OSX10.12 x64 Checked Build and Test please

sepidehMS commented:
Ready for review.

AtsushiKan commented on src/mscorlib/src/System/Reflection/Runtime/Assemblies/AssemblyNameHelpers.cs:
```diff
@@ -0,0 +1,156 @@
+using System.Text;
+using System.Collections.Generic;
+
+namespace System.Reflection.Runtime.Assemblies
```
We should move these to the System.Reflection namespace. (Ideally that should have been done last year when these files were moved out of Reflection.Core.)

AtsushiKan commented on src/mscorlib/src/System/Reflection/Runtime/Assemblies/AssemblyNameHelpers.cs:
```diff
@@ -0,0 +1,156 @@
+                if (canonicalizedVersion.Major != ushort.MaxValue)
+                {
+                    sb.Append(", Version=");
+                    sb.Append(a.Version.ToString());
```
You need to continue using the canonicalizedVersion to be compatible with the desktop.

The canonicalizedVersion is what truncates the 32-bit Version components into unsigned 16 bit numbers (to match the underlying metadata structures on both N and ECMA and ensure that the AssemblyName is roundtrippable from them.)

A Version component whose lower 16-bits is 0xffff is treated as the "not specified" value and causes it and the numbers that come after it to be dropped from the output (see code sample below) It's probably better not to use Version.ToString() here and implement the logic yourself.

```
        public static void Main()
        {
            AssemblyName a = new AssemblyName("Foo");
            a.Version = new Version(1, 2, 0x1ffff, 4);
            Console.WriteLine("Version.ToString()    = " + a.Version);
            Console.WriteLine("AssemblyName.FullName = " + a.FullName);
        }

```
outputs:
 
```
 Version.ToString()    = 1.2.131071.4
 AssemblyName.FullName = Foo, Version=1.2

```

on the Desktop.

sepidehMS replied:
@AtsushiKan Thanks for pointing that out. Fixed.
